src/WP_CLI/RecursiveDataStructureTraverser.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+namespace WP_CLI;
+
+class RecursiveDataStructureTraverser {
+
+	/**
+	 * @var mixed The data to traverse set by reference.
+	 */
+	protected $data;
+
+	/**
+	 * @var string The character/sequence used to delineate hierarchy in a single key.
+	 */
+	protected $delimiter;
+
+	/**
+	 * @var null|string The key the data belongs to in the parent's data.
+	 */
+	protected $key;
+
+	/**
+	 * @var null|static The parent instance of the traverser.
+	 */
+	protected $parent;
+
+	/**
+	 * RecursiveDataStructureTraverser constructor.
+	 *
+	 * @param        $data
+	 * @param static $parent
+	 */
+	public function __construct( &$data, $parent = null ) {
+		$this->data =& $data;
+		$this->parent = $parent;
+	}
+
+	/**
+	 * @param $locator
```

Comment:
Yeah, I was going to add this stuff. This has been an interactive work in progress. I didn't intend for this to be critiqued as ready for merge yet 🤠